File: src/mvmctl/cli/volume.py
"""Volume management commands."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any

import typer

from mvmctl.api import ConfigOperation as _ConfigOperation
from mvmctl.api import VolumeCreateInput as _VolumeCreateInput
from mvmctl.api import VolumeInput as _VolumeInput
from mvmctl.api import VolumeOperation as _VolumeOperation
from mvmctl.cli._completion import _complete_volume_names
from mvmctl.utils.cli import handle_errors, mvm_cli

if TYPE_CHECKING:
    from mvmctl.api.config_operations import ConfigOperation
    from mvmctl.api.inputs._volume_create_input import VolumeCreateInput
    from mvmctl.api.inputs._volume_input import VolumeInput
    from mvmctl.api.volume_operations import VolumeOperation
else:
    ConfigOperation = _ConfigOperation
    VolumeOperation = _VolumeOperation
    VolumeInput = _VolumeInput
    VolumeCreateInput = _VolumeCreateInput
from mvmctl.cli._common import (
    ListingColumn,
    render_listing,
    resolve_listing_style,
)

volume_app = typer.Typer(
    help="Volume management",
    no_args_is_help=True,
    add_completion=False,
)


@volume_app.callback()
def volume_callback(ctx: typer.Context) -> None:
    pass


@volume_app.command(name="create")
@handle_errors
def volume_create(
    name: str = typer.Argument(..., help="Volume name"),
    size: str = typer.Argument(..., help="Volume size (e.g., 1G, 512M)"),
    format: str | None = typer.Option(
        None, "--format", help="Disk format: raw or qcow2 (default: raw)"
    ),
    read_only: bool | None = typer.Option(
        None,
        "--read-only",
        "--readonly",
        help="Mount volume as read-only (default: writable)",
        is_flag=True,
    ),
) -> None:
    """Create a new persistent volume."""
    result = VolumeOperation.create(
        VolumeCreateInput(
            name=name, size=size, format=format, read_only=read_only
        )
    )
    if result.is_error:
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)
    mvm_cli.success(result.message)
    if result.item:
        mvm_cli.key_value("ID", mvm_cli.format_id(result.item.id))
        mvm_cli.key_value("Mode", "ro" if result.item.is_read_only else "rw")
        mvm_cli.key_value("Format", result.item.format)
        mvm_cli.key_value(
            "Size",
            mvm_cli.format_size(result.item.size_bytes),
        )


@volume_app.command(name="rm")
@handle_errors
def volume_rm(
    identifiers: list[str] = typer.Argument(
        ...,
        help="Volume names or ID prefixes to remove",
        autocompletion=_complete_volume_names,
    ),
    force: bool = typer.Option(
        False, "--force", "-f", help="Remove even if attached to VMs"
    ),
) -> None:
    """Remove one or more volumes."""
    result = VolumeOperation.remove(
        VolumeInput(identifiers=identifiers), force=force
    )
    for r in result.items:
        item_name = r.item.name if r.item else "unknown"
        if r.is_ok:
            mvm_cli.success(f"Removed: {item_name}")
        else:
            mvm_cli.error(r.message or f"Remove failed: {item_name}")
    if result.has_any_error:
        raise typer.Exit(code=1)


_VOLUME_COLUMNS = [
    ListingColumn("ID", lambda v: mvm_cli.format_id(v.id)),
    ListingColumn("Name", lambda v: v.name),
    ListingColumn("Size", lambda v: mvm_cli.format_size(v.size_bytes)),
    ListingColumn("Status", lambda v: v.status),
    ListingColumn("Format", lambda v: v.format, long_only=True),
    ListingColumn("Attached To", lambda v: v.vm_id or "-", long_only=True),
    ListingColumn("Created", lambda v: mvm_cli.format_timestamp(v.created_at)),
]


@volume_app.command(name="ls")
@handle_errors
def volume_ls(
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
    long_output: bool = typer.Option(
        False, "--long", help="Show full listing with all columns"
    ),
) -> None:
    """List all volumes."""
    volumes = VolumeOperation.list_all()

    if json_output:
        data: list[dict[str, Any]] = []
        for vol in volumes:
            data.append(
                {
                    "id": vol.id,
                    "name": vol.name,
                    "size": vol.size_bytes,
                    "size_bytes": vol.size_bytes,
                    "format": vol.format,
                    "is_read_only": vol.is_read_only,
                    "status": vol.status,
                    "vm_id": vol.vm_id,
                    "created_at": vol.created_at,
                }
            )
        typer.echo(json.dumps(data, indent=2))
        return

    style = resolve_listing_style(long_output)

    render_listing(volumes, _VOLUME_COLUMNS, style)


@volume_app.command(name="inspect")
@handle_errors
def volume_inspect(
    identifier: str = typer.Argument(
        ...,
        help="Volume name or ID prefix",
        autocompletion=_complete_volume_names,
    ),
    json_output: bool = typer.Option(False, "--json", help="Output as JSON"),
) -> None:
    """Show detailed information about a volume."""
    info = VolumeOperation.inspect(VolumeInput(identifiers=[identifier]))

    if json_output:
        typer.echo(json.dumps(info, indent=2, default=str))
        return

    mvm_cli.print_dict_tree(info, title=f"Volume: {info['volume']['name']}")


@volume_app.command(name="resize")
@handle_errors
def volume_resize(
    identifier: str = typer.Argument(
        ...,
        help="Volume name or ID prefix",
        autocompletion=_complete_volume_names,
    ),
    size: str = typer.Argument(..., help="New size (e.g., 1G, 512M)"),
) -> None:
    """Resize a volume."""
    result = VolumeOperation.resize(
        VolumeCreateInput(name=identifier, size=size)
    )
    if result.is_error:
        mvm_cli.error(result.message)
        raise typer.Exit(code=1)
    mvm_cli.success(result.message)

File: internal/cli/volume.go
// Package cli — volume management commands
package cli

import (
	"encoding/json"
	"fmt"

	"mvmctl/internal/cli/common"
	"mvmctl/internal/lib/model"
	"mvmctl/pkg/api"
	"mvmctl/pkg/api/inputs"

	"github.com/spf13/cobra"
)

// volumeColumns defines the listing columns for volumes.
var volumeColumns = []common.ListingColumn{
	{Header: "ID", Extract: func(v any) string { return common.Cli.FormatID(v.(*model.VolumeItem).ID) }},
	{Header: "Name", Extract: func(v any) string { return v.(*model.VolumeItem).Name }},
	{Header: "Size", Extract: func(v any) string { return common.Cli.FormatSize(v.(*model.VolumeItem).SizeBytes) }},
	{Header: "Status", Extract: func(v any) string { return string(v.(*model.VolumeItem).Status) }},
	{Header: "Format", Extract: func(v any) string { return string(v.(*model.VolumeItem).Format) }, LongOnly: true},
	{Header: "Attached To", Extract: func(v any) string {
		vmID := v.(*model.VolumeItem).VMID
		if vmID != nil && *vmID != "" {
			return *vmID
		}
		return "-"
	}, LongOnly: true},
	{
		Header:  "Created",
		Extract: func(v any) string { return common.Cli.FormatTimestamp(v.(*model.VolumeItem).CreatedAt, "relative") },
	},
}

func NewVolumeCmd(volumeAPI api.VolumeAPI, configAPI api.ConfigAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:     "volume",
		Aliases: []string{"vol"},
		Short:   "Volume management",
		Long:    "Manage persistent volumes — list, create, remove, inspect, resize, attach, detach.",
	}

	cmd.AddCommand(newVolumeListCmd(volumeAPI, configAPI))
	cmd.AddCommand(newVolumeCreateCmd(volumeAPI))
	cmd.AddCommand(newVolumeRemoveCmd(volumeAPI))
	cmd.AddCommand(newVolumeInspectCmd(volumeAPI))
	cmd.AddCommand(newVolumeResizeCmd(volumeAPI))
	cmd.AddCommand(newVolumeAttachCmd(volumeAPI))
	cmd.AddCommand(newVolumeDetachCmd(volumeAPI))

	return cmd
}

func newVolumeListCmd(volumeAPI api.VolumeAPI, configAPI api.ConfigAPI) *cobra.Command {
	var jsonOutput bool
	var longOutput bool

	cmd := &cobra.Command{
		Use:     "ls",
		Aliases: []string{"list"},
		Short:   "List all volumes",
		Args:    cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			volumes := volumeAPI.VolumeListAll(cmd.Context())

			if jsonOutput {
				if volumes == nil {
					volumes = []*model.VolumeItem{}
				}
				jsonBytes, _ := json.MarshalIndent(volumes, "", "  ")
				fmt.Println(string(jsonBytes))
				return nil
			}

			style := common.Cli.ResolveListingStyle(cmd.Context(), configAPI, longOutput)

			common.RenderListing(volumes, volumeColumns, style)
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	cmd.Flags().BoolVar(&longOutput, "long", false, "Show full listing with all columns")
	return cmd
}

func newVolumeCreateCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	var format string
	var readOnly bool
	var shareable bool

	cmd := &cobra.Command{
		Use:   "create [name] [size]",
		Short: "Create a new persistent volume",
		Args:  cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			name := args[0]
			sizeStr := args[1]

			var formatPtr *string
			if cmd.Flags().Changed("format") {
				formatPtr = &format
			}
			var readOnlyPtr *bool
			if cmd.Flags().Changed("read-only") || cmd.Flags().Changed("readonly") {
				readOnlyPtr = &readOnly
			}
			var shareablePtr *bool
			if cmd.Flags().Changed("shareable") || cmd.Flags().Changed("s") {
				shareablePtr = &shareable
			}
			input := inputs.VolumeCreateInput{
				Name:      name,
				Size:      sizeStr,
				Format:    formatPtr,
				ReadOnly:  readOnlyPtr,
				Shareable: shareablePtr,
			}
			vol, err := volumeAPI.VolumeCreate(cmd.Context(), input)
			if err != nil {
				return err
			}
			common.Cli.Success(fmt.Sprintf("Volume '%s' created", name))
			for _, col := range volumeColumns {
				switch col.Header {
				case "ID", "Format", "Size":
					common.Cli.KeyValue(col.Header, col.Extract(vol), 2, 12)
				}
			}
			mode := "rw"
			if vol.IsReadOnly {
				mode = "ro"
			}
			common.Cli.KeyValue("Mode", mode, 2, 12)
			if vol.IsShareable {
				common.Cli.KeyValue("Shareable", "yes", 2, 12)
			}
			return nil
		},
	}

	cmd.Flags().StringVar(&format, "format", "", "Disk format: raw or qcow2 (default: raw)")
	// Accept both --read-only and --readonly as aliases
	cmd.Flags().BoolVar(&readOnly, "read-only", false, "Mount volume as read-only (default: writable)")
	cmd.Flags().BoolVar(&readOnly, "readonly", false, "Mount volume as read-only (default: writable)")
	cmd.Flags().BoolVar(&readOnly, "ro", false, "Mount volume as read-only (default: writable)")
	cmd.Flags().
		BoolVarP(&shareable, "shareable", "s", false, "Allow volume to be attached to multiple VMs (requires --read-only)")
	return cmd
}

func newVolumeRemoveCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	var force bool

	cmd := &cobra.Command{
		Use:               "rm [identifiers...]",
		Aliases:           []string{"remove", "delete", "del"},
		Short:             "Remove one or more volumes",
		Args:              cobra.MinimumNArgs(1),
		ValidArgsFunction: completeVolumeNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			removeResult := volumeAPI.VolumeRemove(cmd.Context(), inputs.VolumeInput{Identifiers: args}, force)
			for _, r := range removeResult.Items {
				itemName := "unknown"
				if r.Item != nil {
					if vol, ok := r.Item.(*model.VolumeItem); ok {
						itemName = vol.Name
					}
				}
				if r.IsOK() {
					common.Cli.Success(fmt.Sprintf("Removed: %s", itemName))
				} else {
					msg := r.Message
					if msg == "" {
						msg = fmt.Sprintf("Remove failed: %s", itemName)
					}
					common.Cli.Error(msg)
				}
			}
			if removeResult.HasErrors() {
				return fmt.Errorf("one or more removals failed")
			}
			return nil
		},
	}

	cmd.Flags().BoolVarP(&force, "force", "f", false, "Remove even if attached to VMs")
	return cmd
}

func newVolumeInspectCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:               "inspect [identifier]",
		Short:             "Show detailed information about a volume",
		Args:              cobra.ExactArgs(1),
		ValidArgsFunction: completeVolumeNames,
		RunE: func(cmd *cobra.Command, args []string) error {
			identifier := args[0]

			info, err := volumeAPI.VolumeInspect(cmd.Context(), inputs.VolumeInput{Identifiers: []string{identifier}})
			if err != nil {
				return err
			}

			if jsonOutput {
				fmt.Println(common.MarshalJSONDefaultStr(info))
				return nil
			}

			common.Cli.PrintDictTree(common.Cli.ToMap(info), fmt.Sprintf("Volume: %s", info.Volume.Name))
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output as JSON")
	return cmd
}

func newVolumeResizeCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	cmd := &cobra.Command{
		Use:               "resize [identifier] [size]",
		Short:             "Resize a volume",
		Args:              cobra.ExactArgs(2),
		ValidArgsFunction: completeVolumeThenSize,
		RunE: func(cmd *cobra.Command, args []string) error {
			identifier := args[0]
			sizeArg := args[1]

			if err := volumeAPI.VolumeResize(
				cmd.Context(),
				inputs.VolumeCreateInput{Name: identifier, Size: sizeArg},
			); err != nil {
				return err
			}
			common.Cli.Success(fmt.Sprintf("Volume '%s' resized", identifier))
			return nil
		},
	}

	return cmd
}

func newVolumeAttachCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	return &cobra.Command{
		Use:   "attach [vm_identifier] [volume_identifier]",
		Short: "Attach a volume to a VM.",
		Long: `Attach a volume to a running VM.

Arguments:
  vm_identifier      VM identifier (name, ID prefix, IP, or MAC)
  volume_identifier  Name or ID of the volume to attach`,
		Args:              cobra.ExactArgs(2),
		ValidArgsFunction: completeVMThenVolume,
		RunE: func(cmd *cobra.Command, args []string) error {
			id := args[0]
			volumeName := args[1]

			input := inputs.VolumeInput{
				VMIdentifier: id,
				Identifiers:  []string{volumeName},
			}
			if err := volumeAPI.VolumeAttach(cmd.Context(), input); err != nil {
				return fmt.Errorf("attach volume %q: %w", volumeName, err)
			}

			common.Cli.Success(fmt.Sprintf("Volume '%s' attached", volumeName))
			return nil
		},
	}
}

func newVolumeDetachCmd(volumeAPI api.VolumeAPI) *cobra.Command {
	return &cobra.Command{
		Use:   "detach [vm_identifier] [volume_identifier]",
		Short: "Detach a volume from a VM.",
		Long: `Detach a volume from a running VM.

Arguments:
  vm_identifier      VM identifier (name, ID prefix, IP, or MAC)
  volume_identifier  Name or ID of the volume to detach`,
		Args:              cobra.ExactArgs(2),
		ValidArgsFunction: completeVMThenVolume,
		RunE: func(cmd *cobra.Command, args []string) error {
			id := args[0]
			volumeName := args[1]

			input := inputs.VolumeInput{
				VMIdentifier: id,
				Identifiers:  []string{volumeName},
			}
			if err := volumeAPI.VolumeDetach(cmd.Context(), input); err != nil {
				return fmt.Errorf("detach volume %q: %w", volumeName, err)
			}

			common.Cli.Success(fmt.Sprintf("Volume '%s' detached", volumeName))
			return nil
		},
	}
}
